src/core/Akka.Streams/Implementation/Fusing/Ops.cs

```diff
@@ -2513,6 +2515,360 @@ public Expand(Func<TIn, IEnumerator<TOut>> extrapolate)
+                }
+
+                private void VTCompletionError(ValueTask<T> vtCapture)
+                {
+                    var t = vtCapture.AsTask();
+                    //We only care about faulted, not canceled.
+                    if (t.IsFaulted)
+                    {
+                        var exception = t.Exception?.InnerExceptions != null &&
+                                        t.Exception.InnerExceptions.Count == 1
+                            ? t.Exception.InnerExceptions[0]
+                            : t.Exception;
+
+                        Invoke(Result.Failure<T>(exception));
+                    }
+                }
+
+                public Holder(Result<T> element, Action<Holder<T>> callback)
+                {
+                    _callback = callback;
+                    Element = element;
+                }
+
+                public void SetElement(Result<T> result)
+                {
+                    Element = result.IsSuccess && result.Value == null
+                        ? Result.Failure<T>(ReactiveStreamsCompliance
+                            .ElementMustNotBeNullException)
+                        : result;
+                }
+
+                public void SetContinuation(ValueTask<T> vt)
+                {
+                    var valueTask = vt;
+                    var peeker =
+                        Unsafe.As<ValueTask<T>, ValueTaskCheatingPeeker<T>>(
+                            ref valueTask);
+                    if (peeker._obj == null)
+                    {
+                        Invoke(Result.Success(peeker._result));
+                    }
+                    else if (peeker._obj is Task<T> asTask)
+                    {
+                        asTask.ContinueWith(TaskCompletedAction, this,
+                            TaskContinuationOptions.NotOnCanceled);
+                    }
+                    else
+                    {
+                        _pending = vt;
+                        var source =
+                            Unsafe.As<IValueTaskSource<T>>(peeker._obj);
+                        source.OnCompleted(OnCompletedAction, this,
+                            peeker._token,
+                            ValueTaskSourceOnCompletedFlags.None);
+                    }
+                }
+
+                public void Invoke(Result<T> result)
+                {
+                    SetElement(result);
+                    _callback(this);
+                }
+            }
+
+            private static readonly Result<TOut> NotYetThere =
+                Result.Failure<TOut>(NotYetThereSentinel.Instance);
+
+            private readonly SelectValueTaskAsync<TIn, TOut> _stage;
+            private readonly Decider _decider;
+            private IBuffer<Holder<TOut>> _buffer;
+            private readonly Action<Holder<TOut>> _taskCallback;
+
+            private readonly
+                ConcurrentQueue<
+                    Holder<TOut>> _queue;
```

Comment:
The idea here, is that we treat used Holders as 'Lazy' instances, so this will only grow to the size of `parallelism` at the most, and even then only if we get to that level of demand.